### PR TITLE
OSSM-4255: Sync with kubernetes/code-generator v0.27

### PR DESCRIPTION
**Pull request by jewertow**
This PR aligns informer templates with `kubernetes/code-generator` v0.27 which slightly changed in compare to v0.25.1.

Unfortunately, after recent changes in the code-generator, our informers were no longer implementing upstream informers, because new methods, like `InformerFor`, were added to the `SharedInformerFactory`.
`InformerFor` takes an argument `newFunc internalinterfaces.NewInformerFunc`, which is a different type in xns-informer and in upstream informers, so when I was trying to create an instance of gateway api informer using `NewSharedInformerFactory` from xns-informer, compilation was failing.

To make our informers work, I had to change controller templates so that the functions declare upstream types to return, but in the bodies they return internal types.
```
func New(...) upstream.Interface {
  return internal.Implementation(...)
}
```
I added an argument `--informer-package` to the xns-informer-gen and when it's specified, then upstream types are returned by factory methods, as you can see above. If it's not set, then xns-informer-gen uses only the generated informers (we generate informers without `--informers-package` in maistra/api).

Beside that, I implemented missing methods in `multiNamespaceInformer`.

I imported this branch to maistra/istio and maistra/api and it works fine.

### CODE REVIEW

File: cmd/xns-informer-gen/generators/factory.go [hunk 1]
```diff
@@ -177,39 +184,59 @@ func (f *sharedInformerFactory) SetNamespaces(namespaces []string) {
...
+		return
+	}
+
+	for informerType, informer := range f.informers {
+		if !f.startedInformers[informerType] {
+			f.wg.Add(1)
+			// We need a new variable in each loop iteration,
+			// otherwise the goroutine would use the loop variable
+			// and that keeps changing.
+			informer := informer
```
Comment (rcernich) [inline cmd/xns-informer-gen/generators/factory.go:201]: nit.  This seems a bit odd.  Maybe use a different variable name for the loop local variable, or pass it in as a parameter to the anonymous function.

File: cmd/xns-informer-gen/generators/factory.go [hunk 2]
```diff
@@ -177,39 +184,59 @@ func (f *sharedInformerFactory) SetNamespaces(namespaces []string) {
     f.namespaces.SetNamespaces(namespaces)
 }
 
-// Start initializes all requested informers.
 func (f *sharedInformerFactory) Start(stopCh <-chan struct{}) {
-  f.lock.Lock()
-  defer f.lock.Unlock()
+	f.lock.Lock()
+	defer f.lock.Unlock()
```
Comment (rcernich) [inline cmd/xns-informer-gen/generators/factory.go:189]: What is this protecting?  If it is only protecting shuttingDown, should it be unlocked after the following check?  Should it also be rechecked in the start function in the loop?  I'm not sure what shuttingDown does, as the ShutDown function doesn't appear to do anything  except wait for everything to finish starting.  fwiw, I don't think this is that critical, as it's unlike for somebody to start then immediately call shutdown, especially if shtudown doesn't really do anything.  That said, I didn't look through all of the code around shuttingDown.

File: cmd/xns-informer-gen/generators/factory.go (hunk 1 above)
Comment (jewertow) [reply cmd/xns-informer-gen/generators/factory.go:201]: This is copy-paste from the upstream informer generator, so I think it's better to make as few changes as possible.

File: cmd/xns-informer-gen/generators/factory.go (hunk 2 above)
Comment (jewertow) [reply cmd/xns-informer-gen/generators/factory.go:189]: First of all, this is also copied from the upstream code-generator, so I cannot be 100% sure, but it seems the `lock` prevents concurrent access to `shuttingDown`, `informers`, `startedInformers` and `wg`.

**Review by luksa — APPROVED** [review]

**Pull request merged**